README


This archive contains the distribution Dancer-Session-CHI,
version v0.1.5:

  CHI-based session engine for Dancer

This software is Copyright (c) 2013 by Richard Simões.

This is free software, licensed under:

  The MIT (X11) License




Changes
Revision history for Dancer-Session-CHI

v0.1.5    2013-02-26 14:38:13 America/Chicago
    - Added C<purge> method to give access to CHI's method of the same name.
    - Undefined session cookie on session destruction.
    - Changed license to MIT.

v0.1.4    2012-04-11 15:02:20 America/Chicago
    - Reverted previous fix attempt, implemented actual fix by correcting
      session key mismatch.

v0.1.3    2012-04-11 12:51:19 America/Chicago
    - Unblessed session object before passing to CHI to prevent hilarity during
      serialization by Storable.

v0.1.2    2012-04-11 09:26:18 America/Chicago
    - Module fully reimplemented without Moose.
    - Fixed issue where CHI, when using file-based cache systems, would place
      the session directory in $app_dir/public if given a relative path.

v0.1.1    2011-11-08 13:03:37 CST6CDT
    - Remove unused Scalar::Util prereq
    - Avoid Dancer's pernicious global actions by removing `use Dancer qw(...);`
    - Refactor tests

v0.1.0    2011-11-06 13:11:29 America/Chicago
    - Change license to LGPL, v3
    - First non-point release

v0.0.5    2011-11-06 12:33:21 America/Chicago
    - Build with Module::Build

v0.0.4    2011-11-06 12:15:20 America/Chicago
    - Removed a couple MooseX:: dependencies and Perl v5.10-related pragmas

v0.0.3    2011-10-24 05:03:13 America/Chicago
    - Removed calls to Dancer v1 deep guts in anticipation of changes in Dancer v2.

v0.0.2    2011-10-24 01:23:18 America/Chicago
    - Avoided smart-match bug in Perl <= v5.12 by not using "given".
    - Set version requirements for most dependencies

v0.0.1    2011-10-23 03:43:17 America/Chicago
    - First version, released on an unsuspecting world.


Makefile.PL

use strict;
use warnings;

use 5.008;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "CHI-based session engine for Dancer",
  "AUTHOR" => "Richard Sim\303\265es <rsimoes at cpan dot org>",
  "BUILD_REQUIRES" => {
    "Carp::Always" => "0.10",
    "Dancer::Test" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Test::More" => 0,
    "Test::Most" => "0.22",
    "strict" => 0,
    "warnings" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Dancer-Session-CHI",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "NAME" => "Dancer::Session::CHI",
  "PREREQ_PM" => {
    "CHI" => "0.49",
    "Dancer" => "1.3072",
    "Dancer::Plugin::Cache::CHI" => 0,
    "Dancer::Session::Abstract" => 0,
    "English" => 0,
    "Moose" => "2.0205",
    "MooseX::ClassAttribute" => "0.26",
    "MooseX::NonMoose" => "0.22",
    "Scalar::Util" => 0,
    "namespace::autoclean" => "0.13",
    "strictures" => 1,
    "utf8" => 0
  },
  "VERSION" => "v0.0.4",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Dancer/Session/CHI.pm
package Dancer::Session::CHI;

use strict;
use warnings;
use utf8;
use Carp;
use CHI;
use Dancer qw(cookie);
use Dancer::Logger;
use Dancer::Config qw(setting);
use Dancer::ModuleLoader;
use File::Spec::Functions qw(rel2abs);
use Scalar::Util qw(blessed);

use base "Dancer::Session::Abstract";

our $VERSION = 'v0.1.5'; # TRIAL VERSION
# ABSTRACT: CHI-based session engine for Dancer

# Class methods:

sub create {
    my ($class) = @_;
    my $self = $class->new;
    $self->flush;
    Dancer::Logger->debug("Session (id: " . $self->id . " created.");
    return $self;
}

sub retrieve {
    my (undef, $session_id) = @_;
    my $chi = _build_chi();
    return $chi->get("session_$session_id")
}

# Object methods:

sub flush {
    my ($self) = @_;
    my $chi = _build_chi();
    my $session_id = $self->id;
    $chi->set( "session_$session_id" => $self );
    return $self;
}

sub purge {
    my ($self) = @_;
    my $chi = _build_chi();
    $chi->purge;
    return $self;
}

sub destroy {
    my ($self) = @_;
    my $session_id = $self->id;
    my $chi = _build_chi();
    $chi->remove("session_$session_id");
    cookie setting("session_name") => undef;
    Dancer::Logger->debug("Session (id: $session_id) destroyed.");
    return $self;
}

my $chi;
sub _build_chi {

    return $chi if blessed($chi) && $chi->isa("CHI");

    my $options = setting("session_CHI");
    ( ref $options eq ref {} ) or croak "CHI session options not found";

    # Don't let CHI determine the absolute path:
    if ( exists $options->{root_dir} ) {
        $options->{root_dir} = rel2abs($options->{root_dir});
    }

    my $use_plugin = delete $options->{use_plugin};
    my $is_loaded = exists setting("plugins")->{"Cache::CHI"};
    if ( $use_plugin && !$is_loaded ) {
        croak "CHI plugin requested but not loaded";
    }

    return $use_plugin
        ? do {
            my $plugin = "Dancer::Plugin::Cache::CHI";
            my $error_msg = "$plugin is needed and is not installed";
            unless ( Dancer::ModuleLoader->load($plugin) ) {
                raise( core_session => $error_msg );
            }
            Dancer::Plugin::Cache::CHI::cache()
        }
        : CHI->new( %{$options} );
}

1;
=encoding utf8

=head1 NAME

Dancer::Session::CHI - CHI-based session engine for Dancer

=head1 SYNOPSIS

In a L<Dancer> application:

    set session          => "CHI";
    set session_expires  => "1 hour";
    set session_CHI      => { use_plugin => 1 };

    set plugins          => {
        "Cache::CHI" => {
            driver => 'Memory',
            global => 1
        }
    };

In a F<config.yml>:

    session: CHI
    session_expires: 1 hour
    session_CHI:
        use_plugin: 1

    plugins:
        Cache::CHI:
            driver: Memory
            global: 1

=head1 DESCRIPTION

This module leverages L<CHI> to provide session management for L<Dancer>
applications. Just as L<Dancer::Session::KiokuDB> opens up L<KiokuDB>'s
full range of C<KiokuDB::Backend>::* modules to be used in Dancer session
management, L<Dancer::Session::CHI> makes available the complete
C<CHI::Driver>::* collection.

=head1 CONFIGURATION

Under its C<session_CHI> key, Dancer::Session::CHI accepts a C<use_plugin>
option that defaults to C<0>. If set to C<1>, L<Dancer::Plugin::Cache::CHI>
will be used directly for session management, with no changes made to the
plugin's configuration.

If C<use_plugin> is left false, all other options are passed through to
construct a new L<CHI> object, even if L<Dancer::Plugin::Cache::CHI> is also in
use. This new object needn't use the same L<CHI::Driver> as the plugin.

=head1 METHODS

=for Pod::Coverage BUILD

=head2 CLASS

=over

=item C<create()>

Creates a new session object and returns it.

=item C<retrieve($id)>

Returns the session object containing an ID of $id.

=back

=head2 OBJECT

=over

=item C<flush()>

Writes all session data to the CHI storage backend.

=item C<destroy()>

Ends a Dancer session and wipes the session's data from the CHI storage backend.

=item C<purge()>

Direct access to CHI's C<purge()> method, clearing the data of all expired
sessions from the CHI storage backend.

=back

=head1 CAVEATS

=over

=item *

Some L<CHI::Driver> parameters are sufficiently complex to not be placeable in a F<config.yml>. Session and/or
plugin configuration may instead be needed to be done in application code.

=item *

When using L<CHI::Driver::DBI>, thread/fork safety can be ensured by passing it
a L<DBIx::Connector> object or database handle returned by
L<Dancer::Plugin::Database>'s C<database()> subroutine.

=back

=head1 BUGS

This is an initial I<TRIAL> release, so bugs may be lurking. Please report any issues to this module's
L<GitHub issues page|https://github.com/rsimoes/Dancer-Session-CHI/issues>.

=head1 AUTHOR

Richard Simões <rsimoes at CPAN dot org>

=head1 COPYRIGHT & LICENSE

Copyright © 2013 Richard Simões. This module is released under the terms of the
B<MIT License> and may be modified and/or redistributed under the same or any
compatible license.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Dancer::Plugin::Cache::CHI" => 0,
    "Dancer::Test" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "Module::Build" => "0.3601",
    "Test::Exception" => 0,
    "Test::More" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "CHI-based session engine for Dancer",
  "dist_author" => [
    "Richard Sim\303\265es <rsimoes AT cpan DOT org>"
  ],
  "dist_name" => "Dancer-Session-CHI",
  "dist_version" => "v0.1.5",
  "license" => "mit",
  "module_name" => "Dancer::Session::CHI",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "CHI" => 0,
    "Dancer" => "1.3111",
    "Scalar::Util" => 0,
    "perl" => "5.008"
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
